File: Q45_JUMP_GAME/Solution.cpp
//=================2022_12_10===================
// Idea: select the far most position (maxPos) that can jump to, from 
// the intervals of jumpming steps last time
  
class Solution {
public:
    int jump(vector<int>& nums) {
        int maxPos=0;   //the far most position that can jump to
	n=nums.size();
	end=0;	        //right interval of last jump
	step=0;		//jump num

        for(int i=0;i<n-1;++i){
            if(maxPos>=i){

                maxPos=max(maxPos, i+nums[i]);// reach the right interval from the jump steps last time
		
                if(i==end){
                    end=maxPos; //the postion for next jump start
                    ++step;
                }
            }
        }
        return step;
    }
};

//TC:O(N)
//SC:O(1)

File: Q45_JUMP_GAME/Solution.java
class Solution {
    public int jump(int[] nums) {
        int position=nums.length-1;
        int steps=0;
        while(position>0){
            for(int i=0;i<position;i++){
                if(i+nums[i]>=position){
                    position=i;
                    steps++;
                }
            }
        }
        return steps;
    }
}
//TC:O(N^2)
//SC:O(1)